browser/extensions/api/brave_playlists_api.cc

```diff
@@ -0,0 +1,269 @@
+
+namespace CreatePlaylist = extensions::api::brave_playlists::CreatePlaylist;
+namespace GetPlaylist = extensions::api::brave_playlists::GetPlaylist;
+namespace DeletePlaylist = extensions::api::brave_playlists::DeletePlaylist;
+namespace RequestDownload = extensions::api::brave_playlists::RequestDownload;
+
+namespace {
+
+constexpr char kNotInitializedError[] = "Not initialized";
+constexpr char kAlreadyInitializedError[] = "Already initialized";
+constexpr char kInvalidArgsError[] = "Invalid arguments";
+constexpr char kUnknownError[] = "Unknown";
+constexpr char kNotExistPlaylistError[] = "Playlist does not exist";
+constexpr char kFeatureDisabled[] = "Playlist feature is disabled";
+
+PlaylistsController* GetPlaylistsController(content::BrowserContext* context) {
+  brave_playlists::PlaylistsService* service =
+      PlaylistsServiceFactory::GetInstance()
+          ->GetForProfile(Profile::FromBrowserContext(context));
+  return service ? service->controller() : nullptr;
+}
+
+CreatePlaylistParams GetCreatePlaylistParamsFromCreateParams(
+    const CreatePlaylist::Params::CreateParams& params) {
+  CreatePlaylistParams p;
+  p.playlist_name = params.playlist_name;
+  p.playlist_thumbnail_url = params.thumbnail_url;
+
+  for (const auto& file : params.video_media_files)
+    p.video_media_files.emplace_back(file.url, file.title);
+  for (const auto& file : params.audio_media_files)
+    p.audio_media_files.emplace_back(file.url, file.title);
+  return p;
+}
+
+}  // namespace
+
+namespace extensions {
+namespace api {
+
+ExtensionFunction::ResponseAction BravePlaylistsCreatePlaylistFunction::Run() {
+  auto* controller = GetPlaylistsController(browser_context());
+  if (!controller) {
+    return RespondNow(Error(kFeatureDisabled));
+  }
+
+  if (!controller->initialized())
+    return RespondNow(Error(kNotInitializedError));
+
+  std::unique_ptr<CreatePlaylist::Params> params(
+      CreatePlaylist::Params::Create(*args_));
+  EXTENSION_FUNCTION_VALIDATE(params.get());
+
+  if (GetPlaylistsController(browser_context())
+          ->CreatePlaylist(
+              GetCreatePlaylistParamsFromCreateParams(params->create_params))) {
+    return RespondNow(NoArguments());
+  }
+
+  return RespondNow(Error(kInvalidArgsError));
+}
+
+ExtensionFunction::ResponseAction BravePlaylistsIsInitializedFunction::Run() {
+  auto* controller = GetPlaylistsController(browser_context());
+  if (!controller) {
+    return RespondNow(Error(kFeatureDisabled));
+  }
+
+  return RespondNow(OneArgument(std::make_unique<base::Value>(
+      controller->initialized())));
+}
+
+ExtensionFunction::ResponseAction BravePlaylistsInitializeFunction::Run() {
+  auto* controller = GetPlaylistsController(browser_context());
+  if (!controller) {
+    return RespondNow(Error(kFeatureDisabled));
+  }
+
+  if (controller->initialized()) {
+    return RespondNow(Error(kAlreadyInitializedError));
+  }
+
+  if (PlaylistsServiceFactory::GetInstance()
```

Comment:
why does this have to be called by the api? Why doesn't it just initialize when it gets created? I would think the only reason to have a separate method is to get an async callback, but it doesn't have one